src/server/stream_family.cc

```diff
@@ -1040,55 +1042,144 @@ void XGroupHelp(CmdArgList args, ConnectionContext* cntx) {
+      trim_args.maxlen = opts.max_len;
+    } else if (opts.trim_strategy == TrimStrategy::kMinId) {
+      trim_args.minid = opts.minid.val;
+    }
+    deleted = streamTrim(s, &trim_args);
+  }
+
+  return deleted;
+}
+
+optional<pair<AddTrimOpts, unsigned>> ParseAddOrTrimArgsOrReply(CmdArgList args,
+                                                                ConnectionContext* cntx,
+                                                                bool is_xadd) {
+  AddTrimOpts opts;
+  opts.key = ArgS(args, 0);
 
+  unsigned id_indx = 1;
   for (; id_indx < args.size(); ++id_indx) {
     ToUpper(&args[id_indx]);
     string_view arg = ArgS(args, id_indx);
-    if (arg == "NOMKSTREAM") {
-      add_opts.no_mkstream = true;
-    } else if (arg == "MAXLEN" || arg == "MINID") {
-      if (arg == "MAXLEN") {
-        add_opts.trim_strategy = TrimStrategy::kAddOptsTrimMaxLen;
-      } else {
-        add_opts.trim_strategy = TrimStrategy::kAddOptsTrimMinId;
+
+    size_t remaining_args = args.size() - id_indx - 1;
+
+    if (is_xadd && arg == "NOMKSTREAM") {
+      opts.no_mkstream = true;
+    } else if (arg == "MAXLEN" && remaining_args >= 1) {
+      if (opts.trim_strategy != TrimStrategy::kNone) {
+        (*cntx)->SendError("MAXLEN and MINID options at the same time are not compatible",
+                           kSyntaxErr);
+        return std::nullopt;
       }
-      if (id_indx + 2 >= args.size()) {
-        return (*cntx)->SendError(kSyntaxErr);
+
+      id_indx++;
+      arg = ArgS(args, id_indx);
+      if (remaining_args >= 2 && arg == "~") {
+        opts.trim_approx = true;
+        id_indx++;
+        arg = ArgS(args, id_indx);
+      } else if (remaining_args >= 2 && arg == "=") {
+        opts.trim_approx = false;
+        id_indx++;
+        arg = ArgS(args, id_indx);
       }
-      ++id_indx;
-      if (ArgS(args, id_indx) == "~") {
-        add_opts.trim_approx = true;
-        ++id_indx;
+
+      if (!absl::SimpleAtoi(arg, &opts.max_len)) {
+        (*cntx)->SendError(kInvalidIntErr);
+        return std::nullopt;
       }
+
+      opts.trim_strategy = TrimStrategy::kMaxLen;
+    } else if (arg == "MINID" && remaining_args >= 1) {
+      if (opts.trim_strategy != TrimStrategy::kNone) {
+        (*cntx)->SendError("MAXLEN and MINID options at the same time are not compatible",
+                           kSyntaxErr);
+        return std::nullopt;
+      }
+
+      id_indx++;
       arg = ArgS(args, id_indx);
-      if (add_opts.trim_strategy == TrimStrategy::kAddOptsTrimMaxLen &&
-          !absl::SimpleAtoi(arg, &add_opts.max_len)) {
-        return (*cntx)->SendError(kSyntaxErr);
+      if (remaining_args >= 2 && arg == "~") {
+        opts.trim_approx = true;
+        id_indx++;
+        arg = ArgS(args, id_indx);
+      } else if (remaining_args >= 2 && arg == "=") {
+        opts.trim_approx = false;
+        id_indx++;
+        arg = ArgS(args, id_indx);
       }
```

Comment:
Its the same as above, except for the three lines below. I think we can handle MAXLEN and MINID within one branch by differentiating those two cases only at the very end